tests/blackbox_asserts.py

```diff
@@ -0,0 +1,298 @@
+                    DRProp.passed: lambda args: bool(math.gcd(*args)),
+                    DRProp.rejected: lambda args: not bool(math.gcd(*args)),
+                    DRProp.errorMessage: None,
+                },
+            }
+
+            exec_mode = mode_to_execution_mode(Mode.Application)
+            inputs, assertions = SequenceAssertion.inputs_and_assertions(scenario, exec_mode)
+            euclid_results = Executor.dryrun_app_on_sequence(algod, euclid_app_teal, inputs)
+            for assert_type, predicate in assertions.items():
+                assertion = SequenceAssertion(predicate, name=str(assert_type))
+                assertion.dryrun_assert(inputs, euclid_results, assert_type)
+    """
+    input_types = subr.subroutine.input_types
+    assert (
+        input_types is not None
+    ), "please provide input_types in your @Subroutine annotation (crucial for generating proper end-to-end testable PyTeal)"
+
+    arg_names = subr.subroutine.arguments()
+
+    def arg_prep_n_call(i, p):
+        name = arg_names[i]
+        by_ref = name in subr.subroutine.by_ref_args
+        arg_expr = Txn.application_args[i] if mode == Mode.Application else Arg(i)
+        if p == TealType.uint64:
+            arg_expr = Btoi(arg_expr)
+        prep = None
+        arg_var = arg_expr
+        if by_ref:
+            arg_var = ScratchVar(p)
+            prep = arg_var.store(arg_expr)
+        return prep, arg_var
+
+    def subr_caller():
+        preps_n_calls = [*(arg_prep_n_call(i, p) for i, p in enumerate(input_types))]
+        preps, calls = zip(*preps_n_calls) if preps_n_calls else ([], [])
+        preps = [p for p in preps if p]
+        invocation = subr(*calls)
+        if preps:
+            return Seq(*(preps + [invocation]))
+        return invocation
+
+    def make_return(e):
+        if e.type_of() == TealType.uint64:
+            return e
+        if e.type_of() == TealType.bytes:
+            return Len(e)
+        if e.type_of() == TealType.anytype:
+            x = ScratchVar(TealType.anytype)
+            return Seq(x.store(e), Int(1337))
+        # TealType.anytype:
+        return Seq(e, Int(1337))
+
+    def make_log(e):
+        if e.type_of() == TealType.uint64:
+            return Log(Itob(e))
+        if e.type_of() == TealType.bytes:
+            return Log(e)
+        return Log(Bytes("nada"))
+
+    if mode == Mode.Signature:
+
+        def approval():
+            return make_return(subr_caller())
+
+    else:
+
+        def approval():
+            if subr.subroutine.returnType == TealType.none:
+                result = ScratchVar(TealType.uint64)
+                part1 = [subr_caller(), result.store(Int(1337))]
+            else:
+                result = ScratchVar(subr.subroutine.returnType)
+                part1 = [result.store(subr_caller())]
+
+            part2 = [make_log(result.load()), make_return(result.load())]
+            return Seq(*(part1 + part2))
```

Comment:
Why log or return something on the top of the stack if the return value is already stored in `result`? It seems like you could directly check the contents of that scratch slot without the caveats of the other two methods

Reply:
I'm not sure I'm following your suggestion. Are you proposing not to log `result` or not to save at the top of the stack, or did you find a more efficient way to go about this?

I want to log in addition to leaving at the top of the stack to provide maximum flexibility to the tester and for compatibility with ABI, in case the subroutine encodes as ABI and the reader uses the ABI decoding functionality of an SDK.

Reply:
FYI - this is now in `utils/blackbox.py L345`